infrastructure/src/main/java/org/corfudb/infrastructure/management/FailureDetector.java

```diff
@@ -199,28 +139,69 @@ PollReport pollRound(
-     * If there are no failures, after a few rounds our polling period converges back to initPeriodDuration.
-     *
-     * @param clientRouters     clientRouters
-     * @param failedNodes       list of failed nodes
-     * @param allConnectedNodes all connected nodes
-     */
-    private void tunePollReportTimeouts(
-            Map<String, IClientRouter> clientRouters, Set<String> failedNodes,
-            Set<String> allConnectedNodes) {
-
-        networkStretcher.modifyDecreasedPeriod();
-        tuneRoutersResponseTimeout(
-                clientRouters, allConnectedNodes, networkStretcher.getCurrentPeriod()
-        );
-
-        // Reset the timeout of all the failed nodes to the max value to set a longer
-        // timeout period to detect their response.
-        tuneRoutersResponseTimeout(clientRouters, failedNodes, networkStretcher.getMaxPeriod());
+    static Map<String, Long> getAdjustedResponseTimeouts(Map<String, IClientRouter> routers, long maxSleepBetweenPolls) {
+        return routers.entrySet().stream().map(entry -> {
+            final IClientRouter router = entry.getValue();
+            long currentConnectionTimeout = router.getTimeoutConnect();
+            Preconditions.checkArgument(maxSleepBetweenPolls > currentConnectionTimeout,
+                    "Max sleep between polls: " + maxSleepBetweenPolls +
+                            " should be greater than connection timeout: " + currentConnectionTimeout);
+            long currentResponseTimeout = router.getTimeoutResponse();
+            long overallRouterTimeout = currentConnectionTimeout + currentResponseTimeout;
+            long proposedResponseTimeout = currentResponseTimeout;
+            if (maxSleepBetweenPolls < overallRouterTimeout) {
+                proposedResponseTimeout = maxSleepBetweenPolls - currentConnectionTimeout;
+            }
+            final String endpoint = entry.getKey();
+            log.trace("For {}: connection timeout - {}, response timeout - {}, sleep between polls - {}", endpoint,
+                    currentConnectionTimeout, proposedResponseTimeout, maxSleepBetweenPolls);
+            return Tuple.of(endpoint, proposedResponseTimeout);
+        }).collect(Collectors.toMap(tuple -> tuple.first, tuple -> tuple.second));
     }
 
+
+    PollReport pollRound(long epoch, UUID clusterID, Set<String> allServers, Map<String, IClientRouter> router,
+                         SequencerMetrics sequencerMetrics, ImmutableList<String> layoutUnresponsiveNodes,
+                         FileSystemStats fileSystemStats) {
+        Map<String, Long> timeouts = getAdjustedResponseTimeouts(router, pollConfig.sleepBetweenPolls.toMillis());
+        for (String server : allServers) {
+            final Long adjustedTimeout = timeouts.get(server);
+            router.get(server).setTimeoutResponse(adjustedTimeout);
+        }
+        List<PollReport> reports = new ArrayList<>();
+        for (int i = 0; i < pollConfig.pollRounds; i++) {
```

Comment:
This seems to be polling at the right cadence, but the detection time is still not ~6 seconds. 

What if the remote local views are not refreshed ?
> Duration METRICS_POLL_INTERVAL = Duration.ofSeconds(3);

I stopped one node and measure +10 seconds till reconfiguration. Here are the logs:

```
2023-08-02T00:18:11.570Z | INFO  |              DetectionWorker-0 |        o.c.i.m.FailureDetector | ===>>router timeout  1500
2023-08-02T00:18:11.570Z | INFO  |              DetectionWorker-0 |        o.c.i.m.FailureDetector | ===>>router timeout  1500
2023-08-02T00:18:11.570Z | INFO  |              DetectionWorker-0 |        o.c.i.m.FailureDetector | ===>>router timeout  1500
2023-08-02T00:18:12.071Z | INFO  |              DetectionWorker-0 |        o.c.i.m.FailureDetector | ===>>Poll iteration 0 took 500 millis.
2023-08-02T00:18:12.071Z | INFO  |              DetectionWorker-0 |        o.c.i.m.FailureDetector | ===>> sleep time 1500
2023-08-02T00:18:14.072Z | INFO  |              DetectionWorker-0 |        o.c.i.m.FailureDetector | ===>>Poll iteration 1 took 500 millis.
2023-08-02T00:18:14.072Z | INFO  |              DetectionWorker-0 |        o.c.i.m.FailureDetector | ===>> sleep time 1500
2023-08-02T00:18:16.073Z | INFO  |              DetectionWorker-0 |        o.c.i.m.FailureDetector | ===>>Poll iteration 2 took 500 millis.
2023-08-02T00:18:16.073Z | INFO  |              DetectionWorker-0 |        o.c.i.m.FailureDetector | ===>> sleep time 1500
2023-08-02T00:18:18.570Z | INFO  |              DetectionWorker-2 |        o.c.i.m.FailureDetector | ===>>router timeout  1500
2023-08-02T00:18:18.570Z | INFO  |              DetectionWorker-2 |        o.c.i.m.FailureDetector | ===>>router timeout  1500
2023-08-02T00:18:18.570Z | INFO  |              DetectionWorker-2 |        o.c.i.m.FailureDetector | ===>>router timeout  1500
2023-08-02T00:18:19.298Z | INFO  |              DetectionWorker-2 |        o.c.i.m.FailureDetector | ===>>Poll iteration 0 took 727 millis.
2023-08-02T00:18:19.298Z | INFO  |              DetectionWorker-2 |        o.c.i.m.FailureDetector | ===>> sleep time 1273
2023-08-02T00:18:21.301Z | INFO  |              DetectionWorker-2 |        o.c.i.m.FailureDetector | ===>>Poll iteration 1 took 729 millis.
2023-08-02T00:18:21.301Z | INFO  |              DetectionWorker-2 |        o.c.i.m.FailureDetector | ===>> sleep time 1271
2023-08-02T00:18:23.303Z | INFO  |              DetectionWorker-2 |        o.c.i.m.FailureDetector | ===>>Poll iteration 2 took 731 millis.
2023-08-02T00:18:23.303Z | INFO  |              DetectionWorker-2 |        o.c.i.m.FailureDetector | ===>> sleep time 1269
2023-08-02T00:18:26.075Z | INFO  |                    failAfter-0 |      o.c.r.c.NettyClientRouter | ===>> sendRequestAndGetCompletable: Remove request (Type: REPORT_FAILURE_REQUEST ID: 351) to localhost:9001 due to timeout! Request:version { corfu_source_code_version: 81888530331351 } request_id: 351 priority: HIGH epoch: 50 cluster_id { lsb: -8042818313600600116 msb: 759574279712228859 } client_id { lsb: -7758280800914977836 msb: 4114914633518629641 } ignore_epoch: true  dur 1501
2023-08-02T00:18:27.319Z | INFO  |              DetectionWorker-0 |        o.c.i.m.FailureDetector | ===>>router timeout  1500
2023-08-02T00:18:27.319Z | INFO  |              DetectionWorker-0 |        o.c.i.m.FailureDetector | ===>>router timeout  1500
2023-08-02T00:18:27.319Z | INFO  |              DetectionWorker-0 |        o.c.i.m.FailureDetector | ===>>router timeout  1500
2023-08-02T00:18:27.821Z | INFO  |              DetectionWorker-0 |        o.c.i.m.FailureDetector | ===>>Poll iteration 0 took 501 millis.
2023-08-02T00:18:27.821Z | INFO  |              DetectionWorker-0 |        o.c.i.m.FailureDetector | ===>> sleep time 1499
2023-08-02T00:18:27.832Z | INFO  |                 layoutServer-0 |             o.c.i.LayoutServer | handleCommitLayoutRequest[357]: New layout committed: Layout(layoutServers=[localhost:9000, localhost:9001, localhost:9002], sequencers=[localhost:9002, localhost:9000, localhost:9001], segments=[Layout.LayoutSegment(replicationMode=CHAIN_REPLICATION, start=0, end=-1, stripes=[Layout.LayoutStripe(logServers=[localhost:9002, localhost:9001])])], unresponsiveServers=[localhost:9000], epoch=51, clusterId=0a8a8cc8-72c5-45fb-9062-2b5b40d477cc)
2023-08-02T00:18:29.821Z | INFO  |              DetectionWorker-0 |        o.c.i.m.FailureDetector | ===>>Poll iteration 1 took 501 millis.


```

Reply:
@Maithem 

I adjusted the `METRICS_POLL_INTERVAL` and `POLICY_EXECUTE_INTERVAL` to 500ms. I don't think there is harm in that. 
I wrote this test and measured detection time:

```
@Test
    public void testFD() throws Exception {
        final int PORT_0 = 9000;
        final int PORT_1 = 9001;
        final int PORT_2 = 9002;
        Process corfuServer_1 = runPersistentServer(corfuSingleNodeHost, PORT_0, false);
        Process corfuServer_2 = runPersistentServer(corfuSingleNodeHost, PORT_1, false);
        Process corfuServer_3 = runPersistentServer(corfuSingleNodeHost, PORT_2, false);
        List<Process> corfuServers = Arrays.asList(corfuServer_1, corfuServer_2, corfuServer_3);
        final Layout twoNodeLayout = getLayout(3);
        BootstrapUtil.bootstrap(twoNodeLayout, retries, PARAMETERS.TIMEOUT_SHORT);
        firstRuntime = createDefaultRuntime();
        Predicate<Layout> check = layout -> {
            boolean twoNodeCluster = layout.getAllServers().size() == 3;
            boolean oneSegment = layout.getSegments().size() == 1;

            return twoNodeCluster && oneSegment;
        };
        waitForLayoutChange(check, firstRuntime);

        shutdownCorfuServer(corfuServer_1);
        log.info("SHUTTING DOWN CORFU SERVER");
        Predicate<Layout> check2 = layout -> !layout.getUnresponsiveServers().isEmpty();
        waitForLayoutChange(check2, firstRuntime);
        log.info("NODE IS DOWN. BRING UP. START COUNTDOWN");
        runPersistentServer(corfuSingleNodeHost, PORT_0, false);
        final long start = System.currentTimeMillis();
        Predicate<Layout> check3 = layout -> layout.getUnresponsiveServers().isEmpty();
        waitForLayoutChange(check3, firstRuntime);
        log.info("LAYOUT RESTORED took: {}", System.currentTimeMillis() - start);
        for (Process server : corfuServers) {
            shutdownCorfuServer(server);
        }
    }
 ```
 
 ```
 2023-08-07T23:52:30.204Z | INFO  |            test-main | org.corfudb.integration.StateTransferIT.testFD:143 | NODE IS DOWN. BRING UP. START COUNTDOWN
 2023-08-07T23:52:39.338Z | INFO  |            test-main | org.corfudb.integration.StateTransferIT.testFD:148 | LAYOUT RESTORED took: 9125
 ```
 So we have 9s detection and heal time. I then looked at the logs to see where extra 3s come from:
 
 ```
 2023-08-07T23:52:30.204Z | INFO  |            test-main | org.corfudb.integration.StateTransferIT.testFD:143 | NODE IS DOWN. BRING UP. START COUNTDOWN
 2023-08-07T23:52:31.734Z | INFO  |              DetectionWorker-1 |        o.c.i.m.FailureDetector | Poll report starts
 ```
 (Took `1.5s` to start corfu)
 
 ```
 2023-08-07T23:52:37.755Z | INFO  |              DetectionWorker-1 |      o.c.c.m.m.MicroMeterUtils | Poll report took: 6025 
 ```
 (Took `6s` to start corfu as configured)
 
 ```
 2023-08-07T23:52:37.841Z | INFO  |                   management-0 |         o.c.i.ManagementServer | handleHealFailure[13]: Acquired healing lock. Performing healing for nodes: [localhost:9001]
2023-08-07T23:52:39.269Z | INFO  |                       client-2 |         o.c.i.m.f.HealingAgent | Healing local node successful: {"localNode":"localhost:9001","graph":{"graph":[{"endpoint":"localhost:9000","ty
pe":"CONNECTED","connectivity":{"localhost:9002":"OK","localhost:9001":"OK","localhost:9000":"OK"},"epoch":0},{"endpoint":"localhost:9001","type":"CONNECTED","connectivity":{"localhost:9002":"OK","localhost:9001"
:"OK","localhost:9000":"OK"},"epoch":0},{"endpoint":"localhost:9002","type":"CONNECTED","connectivity":{"localhost:9002":"OK","localhost:9001":"OK","localhost:9000":"OK"},"epoch":0}]},"action":"HEAL","healed":{"e
ndpoint":"localhost:9001","numConnections":3},"layout":["localhost:9000","localhost:9001","localhost:9002"],"unresponsiveNodes":["localhost:9001"],"epoch":2}
2023-08-07T23:52:39.270Z | INFO  |              DetectionWorker-2 |      o.c.c.m.m.MicroMeterUtils | FD task took: 1514 
```
(Took `1.5s` to heal)

```
2023-08-07T23:52:39.338Z | INFO  |            test-main | org.corfudb.integration.StateTransferIT.testFD:148 | LAYOUT RESTORED took: 9125 
```
~ (`1.5s + 6s + 1.5s`)

Reply:
I have an empty 3 node cluster (state transfer should have any impact here) and I'm measuring the sequencer latency, I see +10 seconds for fail-over. For example, the sequencer starts at 9000, moves to 9001 and then 9002, here are the latencies:

```
dur is 13021
dur is 19026
dur is 16021
```

And this is the test:

```java

    @Test
    public void test() {
        CorfuRuntime rt = new CorfuRuntime("localhost:9000,localhost:9001,localhost:9002").connect();
        
        while (true) {
            long start = System.currentTimeMillis();
            rt.getSequencerView().query();
            long dur = System.currentTimeMillis() - start;

            if (dur > 5000) {
                System.out.println("dur is " + dur);
            }
        }
    }

```

Reply:
@Maithem 
I reproed the test, I have not timed the sequencer query yet, but I was shutting down the primary sequencers in a ring.
Here my outputs: 
```
Primary sequencer is localhost:9000
SHUTTING DOWN CORFU SERVER: localhost:9000
NODE localhost:9000 IS DOWN. BRING UP. START COUNTDOWN
 LAYOUT RESTORED run 0 took: 10140
 
 Primary sequencer is localhost:9001
 SHUTTING DOWN CORFU SERVER: localhost:9001
 NODE localhost:9001 IS DOWN. BRING UP. START COUNTDOWN
 LAYOUT RESTORED run 1 took: 9115
 
 Primary sequencer is localhost:9002
 SHUTTING DOWN CORFU SERVER: localhost:9002
 NODE localhost:9002 IS DOWN. BRING UP. START COUNTDOWN
 LAYOUT RESTORED run 2 took: 9596
 
  Primary sequencer is localhost:9000
  SHUTTING DOWN CORFU SERVER: localhost:9000
  NODE localhost:9000 IS DOWN. BRING UP. START COUNTDOWN
  LAYOUT RESTORED run 2 took: 13591
```

I do see sometimes it takes longer, when I investigated I looked into healing node:

```
2023-08-09T19:18:28.585Z | INFO  |              DetectionWorker-1 |      o.c.c.m.m.MicroMeterUtils | Poll report took: 6019
2023-08-09T19:18:28.593Z | INFO  |              DetectionWorker-2 |         o.c.i.m.f.ClusterGraph | Get decision maker: [localhost:9002, localhost:9001, localhost:9000]
2023-08-09T19:18:28.598Z | INFO  |              DetectionWorker-2 |      o.c.c.m.m.MicroMeterUtils | FD task took: 12
2023-08-09T19:18:29.061Z | INFO  |                 CorfuRuntime-0 |      o.c.c.m.m.MicroMeterUtils | Get metrics took: 0
2023-08-09T19:18:29.062Z | INFO  |              DetectionWorker-0 |        o.c.i.m.FailureDetector | Poll report starts
2023-08-09T19:18:35.073Z | INFO  |              DetectionWorker-0 |      o.c.c.m.m.MicroMeterUtils | Poll report took: 6011
2023-08-09T19:18:35.132Z | INFO  |                   management-0 |         o.c.i.ManagementServer | handleHealFailure[21]: request:detector_epoch: 5 healed_node: "localhost:9001"
2023-08-09T19:18:35.133Z | INFO  |                   management-0 |         o.c.i.ManagementServer | handleHealFailure[21]: Acquired healing lock. Performing healing for nodes: [localhost:9001]
2023-08-09T19:18:35.133Z | INFO  |                   management-0 | .m.ReconfigurationEventHandler | Handle healing: [localhost:9001]
```

On some iterations, it seems like it runs the poll report twice because the FD round does not complete with a heal, but each poll report still takes 6 seconds. FD round does not complete because findDecisionMaker() returns empty, which is not part of polling, and looks like some other issue that is not related to removing the NetworkStretcher. I can either look at it in this PR or in another PR - up to you.

Reply:
@Maithem I also timed the sequencer query and I see the overhead:
```
Primary sequencer is localhost:9000
SHUTTING DOWN CORFU SERVER: localhost:9000
Start quering the sequencer
Sequencer query 0 took: 19122

Primary sequencer is localhost:9001
SHUTTING DOWN CORFU SERVER: localhost:9001
Sequencer query 1 took: 19584
```

But this overhead does not come from FD algorithm as detecting + restoring node to the layout takes 9-10 seconds (6s poll + 1.5s startup + the rest is heal) as previously captured.

So there are two more issues at least: client overhead somewhere after the sequencer has been restored to the layout and skipping on the decision maker in some runs, which makes FD run twice.

Reply:
If you just kill the sequencer node, then there is no start up cost, or heal time. So it shouldn't really be reflected.

On the client side, there is a bug where `invalidateLayout` chooses a random node, which can be the dead node, so in that case it will wait for 5 seconds before the timeout selects another random node.